The first implementation of the stochastic physics scheme SPPT for MPAS.

NingWang325 opened the pull request:
This implementation adds pattern initialization and advancement calls to the atmos_core module, and pattern application calls to atm_time_integration module.   It also modifies Makefiles in core_atmosphere/ and its subdirectory dynamics/.

The file Registry.xml is also modified to add variables for patterns in Gaussian and icosahedral grid.

NingWang325 committed:
The first implementation of the stochastic physics scheme SPPT for MPAS.

JeffBeck-NOAA commented on src/core_atmosphere/dynamics/mpas_atm_time_integration.F:
```diff
@@ -720,11 +723,24 @@ subroutine atm_srk3(domain, dt, itimestep, exchange_halo_group)
+      tend_names(2) = "rvcuten"
+      tend_names(3) = "rublten"
+      tend_names(4) = "rvblten"
+      call stochastic_physics_pattern_apply(domain, 4, tend_names, ierr)
```
Are these the total U and V tendencies after all physics calls have been completed?  Why are there two sets of them?

JeffBeck-NOAA replied:
From Ning: Physics are combined in a "convective" and "BL" tendencies within MPAS.  One set of U and V is convective, the others is for the BL.

JeffBeck-NOAA commented on src/core_atmosphere/dynamics/mpas_atm_time_integration.F:
```diff
@@ -720,11 +723,24 @@ subroutine atm_srk3(domain, dt, itimestep, exchange_halo_group)
+! apply random perturbation pattern to the tendency
+      tend_names(1) = "tend_rtheta_physics"
+!      tend_names(2) = "tend_rho_physics"
+      call stochastic_physics_pattern_apply(domain, 1, tend_names, ierr)
```
Tendency perturbations are only applied to "rtheta" because it was found that there are no "rho" tendencies, is that correct?

Also, "rtheta" is actually virtual potential temperature, correct?

JeffBeck-NOAA replied:
From Ning: "rtheta" is modified virtual potential tempeature which has qv computed and used in its calculation.  "rho" is used in computation of "rtheta" and then discarded, so no need to perturb for SPPT.

JeffBeck-NOAA commented on src/core_atmosphere/dynamics/mpas_atm_time_integration.F:
```diff
@@ -31,6 +31,7 @@ module atm_time_integration
    use mpas_atmphys_driver_microphysics
    use mpas_atmphys_todynamics
    use mpas_atmphys_utilities
+   use stochastic_physics_mpas, only : stochastic_physics_pattern_apply
```
We will probably want to call this something like "mpas_atmphys_stoch_physics" to match the names of the other functions here.

willmayfield replied:
Maybe? At this point are we thinking of stochastic physics as an external library or as a part of mpas?

JeffBeck-NOAA replied:
I interpreted "stochastic_physics_mpas" as the function within MPAS that calls stochastic physics.  That's why I was thinking it would make sense for the function name to be consistent.  Obviously stochastic physics itself is part of the UFS repository, not part of MPAS.

JeffBeck-NOAA replied:
@NingWang325, Ning will modify and push commit to PR branch.

JeffBeck-NOAA commented:
> The file Registry.xml is also modified to add variables for patterns in Gaussian and icosahedral grid.

@NingWang325, I don't see the Registry.xml file in this PR.  Does it need to be added?  Also, can you set the default values for new stochastic physics namelist entries to zero (for the decorrelation lengths and times, for example)?  We want to make sure that 1) when a user omits a namelist entry, there is a warning if the field is necessary, and 2) the value of any missing namelist entries should default to zero.

willmayfield commented:
> > The file Registry.xml is also modified to add variables for patterns in Gaussian and icosahedral grid.
> 
> @NingWang325, I don't see the Registry.xml file in this PR. Does it need to be added? Also, can you set the default values for new stochastic physics namelist entries to zero (for the decorrelation lengths and times, for example)? We want to make sure that 1) when a user omits a namelist entry, there is a warning if the field is necessary, and 2) the value of any missing namelist entries should default to zero.

If the magnitude is all set to zero, will it still call the pattern generator/updater? Didn't we discuss whether to add a namelist entry to turn off each scheme? (like do_sppt, etc. in fv3) I don't remember what we decided to do for that.

JeffBeck-NOAA commented:
@NingWang325, I agree with @willmayfield, we want to have a single logical flag to turn each stochastic physics option on/off.  So "do_sppt" should control what happens regardless of what the other specific namelist entries are set to.  All namelist entries should be ignored and the pattern generator/integrator should not be called if "do_sppt = false".

NingWang325 committed:
Changed module name stochastic_physics_mpas to mpas_stochastic_physics.

NingWang325 committed:
Added SPPTINT (user specified pattern advancement interval) to the SPPT implementation.
If SPPTINT = 0 perturbation pattern advances every time step; if SPPTINT = t > 0 pattern advances after t seconds.
